pool/pool.go

```diff
@@ -1071,3 +1081,94 @@ func (p *Pool) statisticMiddleware(nodeKey []byte, endpoint string, method stat.
 		p.statisticCallback(nodeKey, endpoint, method, duration, err)
 	}
 }
+
+type (
+	slidingWindow struct {
+		windowSize int64 // Nanoseconds
+		limit      int64
+
+		// Buckets
+		prevCount *atomic.Int64
+		currCount *atomic.Int64
+
+		windowStart *atomic.Int64 // Nanoseconds
+
+		rotationMu *sync.Mutex
+	}
+)
+
+func newSlidingWindow(windowSize time.Duration, limit int64) *slidingWindow {
+	var windowStart atomic.Int64
+	windowStart.Store(time.Now().UnixNano())
+
+	return &slidingWindow{
+		windowSize:  int64(windowSize),
+		limit:       limit,
+		prevCount:   &atomic.Int64{},
+		currCount:   &atomic.Int64{},
+		windowStart: &windowStart,
+		rotationMu:  &sync.Mutex{},
+	}
+}
+
+// Inc increments the current event counter and manages window rotation.
+func (sw *slidingWindow) Inc() {
+	sw.advance()
+	sw.currCount.Add(1)
+}
+
+// Allow returns true if the errors count is below the threshold.
+func (sw *slidingWindow) Allow() bool {
```

Comment:
You're never using inc/allow separately and a combined implementation can be much more efficient (single `time.Now()`, single `advance()`).